doc/statemachines.tex

```diff
@@ -0,0 +1,310 @@
+
+  Todo: Definition of the terms used in the document:
+  What are the terms that we use to describe the network layer ?
+  Can we reference a common source / textbook ?
+\begin{description}
+\item[Protocol]
+\item[Node]  
+\item[Agent]
+  A unit of execution that follows particular Protocol.
+  (Could be implemented as a green thread for example.)
+  A Node can run many agents.
+\item[Client] The agent that is active in the initial state.
+\item[Server] The agent that waits for a message clients when it starts.  
+\item[Upstream,Downstream,Topology]
+\end{description}
+
+\begin{description}
+\item[State Machine]
+  The Protocols are described as a state machine.
+  Formally the state machine is a tuple $(S,T,initial,final)...$ (States, Transitions, initial state, finial state)...
+  The specification uses different representations of the state machine that all describe the same protocol
+  but may leave out some aspects.
+  \begin{itemize}
+  \item The state machine can be represented as a collection of tables (States, Transitions/Messages..).
+  There is a one to one correspondence between these tables and the Haskell source.
+\item A transition graph of the global view of the state machine.
+\item A transition graph of the client view.
+\item A transition graph of the server view.
+\end{itemize}
+  
+\item[States]
+  The specification assumes a global view of the state of a system consisting of the client and the server.
+  Client and server are always in identical states (by definition)
+  which also means that client and server simultaneously transit to new states.
+  The global state is not observable by the client nor the server.
+  Possible interpretations: 
+  A sequence of states is a set of points in time that a correct run of the protocol passes through.
+  (todo: draw diagram)
+\item[Messages/Events]
+  A tuple $(lable,data)$. Can be serialized transferred over the network and de-serialized.
+  (authenticity of messages not handled here)
+\item[Transitions]
+  The document describes the global/abstract state transitions of the protocol
+  in terms of messages and data that the client and the server exchange.
+  In sync with state transitions the  client and server process the data locally to compute the
+  the actual "outcome" of a protocol-run. (For example "update the block chain").
+  Parts of the local computations are specific to the implementation language and the underlying
+  data structures, but they are also relevant for the protocol (to determine the next transition).
+\item[Initial state]
+\item[Final state \& termination]
+\end{description}
+  
+\subsubsection{Protocol Description}
+
+
+\section{The Chain Synchronization Protocol}
+This section describes the Chain Synchronization Protocol.
+
+\begin{description}
+\item[purpose of the protocol]
+The Chain Synchronization Protocol is used by the block chain consumer
+to synchronize its block chain with the block chain producer.
```

Comment:
It is polymorphic on blocks.  And for some clients it will synchronize just headers (node-to-node), for others it can be used to synchronize actual block (e.g., node-to-wallet).  It might be worth mentioning that, so somebody that implements it can save his work too and not be confused, why there's a separate block download protocol.